protocols/noise/src/xx.rs

```diff
@@ -0,0 +1,198 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::{
+    error::NoiseError,
+    io::NoiseOutput,
+    keys::{Curve25519, PublicKey},
+    util::{to_array, Resolver}
+};
+use futures::prelude::*;
+use snow;
+use std::mem;
+use tokio_io::{AsyncRead, AsyncWrite};
+
+#[derive(Debug, Clone)]
+pub enum XX {}
+
+pub struct NoiseInboundFuture<T>(InboundState<T>);
+
+impl<T> NoiseInboundFuture<T> {
+    pub(super) fn new(io: T, c: super::NoiseConfig<XX>) -> Self {
+        Self(InboundState::Init(io, c))
+    }
+}
+
+enum InboundState<T> {
+    Init(T, super::NoiseConfig<XX>),
+    RecvHandshake1(NoiseOutput<T>), // -> e
+    SendHandshake(NoiseOutput<T>),  // <- e, ee, s, es
+    Flush(NoiseOutput<T>),
+    RecvHandshake2(NoiseOutput<T>), // -> s, se
+    Done
+}
+
+impl<T> Future for NoiseInboundFuture<T>
+where
+    T: AsyncRead + AsyncWrite
+{
+    type Item = (PublicKey<Curve25519>, NoiseOutput<T>);
+    type Error = NoiseError;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        loop {
+            match mem::replace(&mut self.0, InboundState::Done) {
+                InboundState::Init(io, config) => {
+                    let session = snow::Builder::with_resolver(config.params, Box::new(Resolver))
+                        .local_private_key(config.keypair.secret().as_ref())
+                        .build_responder()?;
+                    let output = NoiseOutput::new(io, session);
+                    self.0 = InboundState::RecvHandshake1(output)
+                }
+                InboundState::RecvHandshake1(mut io) => {
+                    if io.poll_read(&mut [])?.is_ready() {
```

Comment:
Every single call to `poll_read` or `poll_write` passes an empty slice.
What's the reason for going for `AsyncRead` and `AsyncWrite` if you're always passing empty slices?

Reply:
During the handshake phase there is no application data to send, so effectively only key material is exchanged. The trace log highlights what is happening. Here is an excerpt for an initial XX handshake write:

```
[TRACE libp2p_noise::io] write state: Init
[TRACE libp2p_noise::io] write state: BufferData { off: 0 }
[TRACE libp2p_noise::io] write: buffered 0 bytes
[TRACE libp2p_noise::io] flush: encrypting 0 bytes
[TRACE libp2p_noise::io] flush: cipher text len = 32 bytes
[TRACE libp2p_noise::io] flush: writing len (32)
[TRACE libp2p_noise::io] flush: wrote 32/32 bytes
[TRACE libp2p_noise::io] flush: finished writing 32 bytes
...
```

Reply:
Oh I see, the `Read` and `Write` are used post-handshake.